lib/broadway/processor.ex

```diff
@@ -112,6 +114,31 @@ defmodule Broadway.Processor do
     :telemetry.execute([:broadway, :processor, :stop], measurements, metadata)
   end
 
+  defp maybe_prepare_messages(messages, state) do
+    %{module: module, context: context, batchers: batchers} = state
+
+    if function_exported?(module, :prepare_messages, 2) and messages != [] do
+      try do
+        messages
+        |> module.prepare_messages(context)
+        |> Enum.map(&validate_message(&1, batchers, "prepare_messages/2"))
+        |> split_by_status([], [])
```

Comment:
I am thinking it is best if `prepare_messages/2` does not allow the messages to fail. All messages returned by `prepare_messages` will be given to `handle_message`. This way we avoid the ambiguity on where work should be done (and it is also less code). :)

Reply:
If messages are not allowed to fail what should be the behaviour in the following cases:
- Invalid messages returned by `prepare_messages/2`
- Exceptions in `prepare_messages/2`

Reply:
Invalid messages are still given to `handle_message`. However, if there is an exception, then we don't call `handle_message` at all, for any of them, yeah.

Reply:
I.e. for exceptions, just keep the code as is!